test/e2e/internal/utils/registry.go

```diff
@@ -0,0 +1,74 @@
+package utils
+
+import (
+	"crypto/rand"
+	"fmt"
+	"net/url"
+	"path"
+
+	oregistry "oras.land/oras-go/v2/registry"
+)
+
+type registry struct {
+	Host      string
+	Username  string
+	Password  string
+	BaseImage string
+}
+
+func (r *registry) Validate() error {
+	if _, err := url.ParseRequestURI(r.Host); err != nil {
+		return err
+	}
+	ref, err := oregistry.ParseReference(r.BaseImage)
+	if err != nil {
+		return err
+	}
+	if ref.Registry != r.Host {
+		return fmt.Errorf("registry host %q mismatch base image %q", r.Host, r.BaseImage)
+	}
+	return nil
+}
+
+func NewRegistry(host, username, password, baseImage string) registry {
+	return registry{
```

Comment:
To follow standard libraries, let's return a pointer of registry, i.e. *registry, and hence return &registry{xxx}